feat: add User Search

babblebey opened the pull request:
## Description

<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->

This Pull Request introduces a User Search feature to enhance the OpenSauced platform's search capabilities. 

### Changes Made

- **SearchDialog Component**: A new `SearchDialog` component has been added. This component is displayed when a user clicks the `SearchDialogTrigger` component, providing an intuitive way to initiate searches.

- **Keyboard Shortcuts**: implemented keyboard shortcuts for quick access to the search functionality. Pressing `Ctrl + K` opens the `SearchDialog`. And, the `Escape` key can be used to close the dialog. 

- **Keyboard Controls**: implemented controls to navigate search results using the `ArrowUp` and `ArrowDown` keys, which activate a cursor trackable by a `_cursorActive` class name within the `UserResultCard` list controlled by the `cursor` state. And pressing `Enter` when a result is selected will navigate to the associated user profile page.

```js
const handleKeyboardCtrl: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
  const resultsCount = searchResult?.data?.length || 0;

  // ArrowUp 🔼 
  if (resultsCount && e.key === "ArrowUp") {
    e.preventDefault();
    setCursor(cursor === 0 ? Math.min(resultsCount - 1, 9) : cursor - 1);
  };

  // ArrowDown 🔽 
  if (resultsCount && e.key === "ArrowDown") {
    e.preventDefault();
    setCursor(cursor === Math.min(resultsCount - 1, 9) ? 0 : cursor + 1);
  };

  // Enter ↩️ 
  if (resultsCount && e.key === "Enter") {
    e.preventDefault();
    if (document.querySelector("._cursorActive")) {
      const user = document.querySelector("._cursorActive") as HTMLAnchorElement;
      router.push(user.href);
      setOpenSearch(false);
    }
  }
};
```

- **Search Info and Error Components**: implemented the `SearchInfo` and `SearchError` components according to the feature design requirement. The `SearchInfo` component provides guidance on the minimum characters required to initiate a search, while the `SearchError` component displays when no results are found, keeping users informed.

![image](https://github.com/open-sauced/app/assets/25631971/11039a7f-aedf-46c7-a7dd-3d767c0fbb3a)

![image](https://github.com/open-sauced/app/assets/25631971/cc4545c3-3502-473c-accb-29c3f79cc104)

- **Search Loading Component**: implemented a `SearchLoading` component to manage transitions between loading and ready states.

![image](https://github.com/open-sauced/app/assets/25631971/388827ea-295d-49a3-ba48-ae7a25016a99)

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

Fixes #1647 

## Mobile & Desktop Screenshots/Recordings

<!-- Visual changes require screenshots -->
[screencast-localhost_3000-2023.09.21-00_17_00.webm](https://github.com/open-sauced/app/assets/25631971/06522ae7-38fe-4f59-ba79-6f65412a681d)


## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

NA

## [optional] What gif best describes this PR or how it makes you feel?



<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

OgDev-01 commented:
Hey @babblebey, The API for this is ready [here](https://beta.api.opensauced.pizza/#/User%20service/getUsersByFilter) FYI 👍

babblebey commented:
> Hey babblebey, The API for this is ready [here](https://beta.api.opensauced.pizza/#/User%20service/getUsersByFilter) FYI 👍 

Great stuff, thanks @OgDev-01

isabensusan commented:
@babblebey  Can we do something like this where we only show the search icon?
<img width="522" alt="image" src="https://github.com/open-sauced/app/assets/27322879/deb93117-3188-44e7-aa35-f757a825d97b">

bdougie commented:
How is this coming along @babblebey. Are you blocked?

babblebey commented:
Oh my apologies! @bdougie 

I'm working on it, No blockers!

Was a little occupied, conclusion shortly 🙏🏾🙏🏾

brandonroberts commented:
@babblebey looks great! The modal doesn't close when I click on a user from the search results though

babblebey commented:
> babblebey looks great! The modal doesn't close when I click on a user from the search results though 

Oh wow! 🤦🏾‍♂️

I can't believe I skipped that, and I considered that in my `***KeyboardCtrls` onClick `Enter` key.


I will address it, thanks for pointing this to me @brandonroberts

bdougie commented:
Took a look on mobile and the logged out view is not quite there.

![image](https://github.com/open-sauced/app/assets/5713670/27be946e-0d5e-4b05-b34f-4dc51b8a67b2)

I also confirmed that I cannot select any users in the list. OnClick doesn't work but when I long press on iOS I can open via the peak screen.

![image](https://github.com/open-sauced/app/assets/5713670/fb0c96bd-5176-442a-9bfc-ff0923797022)

babblebey commented:
> Took a look on mobile and the logged out view is not quite there.

Yea yea! The Arrangement on that `TopNav` component really needs some refactoring! 

I had the SearchTrigger seat inside of one of the components specifically `AuthSection` to hold it in the `flex` there.

I'll appropriate.

> I also confirmed that I cannot select any users in the list. OnClick doesn't work but when I long press on iOS I can open via the peak screen.

It actually does work! It navigates to the User's profile behind the Modal! 

But, the Modal's `openSearch` state is still set to true 😅, exactly what @brandonroberts talked about

babblebey commented:
### New Changes

- **Fix Top Nav:** I refactored the `TopNav` component extracting the `nav` element that wraps the menu items into a Separate component, and reused in a separate `div` optimized for mobile.
- **Move `openSearch` state:** I moved the `openSearch` state to the `GlobalState`'s `store`, this because the component is getting drilled down too many components within the `SearchDialog` and becoming uneasy to manage.
- **Implement Routing in `UserResultCard`:** I implemented `routing` into the `UserResultCard` in place of default `anchor` navigation action; I found that a click on the `anchor` element's navigation to the required route/page is noticeably slower than `routing`; this causes the `SearchDialog` to close (via `onClick` action first) but page still waits before loading the required route/page (I guess the speed varies depending on user's network connection). The usage of `routing` proves faster as its the functionality already implemented with selection via `KeyboardCtrls`.

```js
const handleClick: React.UIEventHandler<HTMLAnchorElement> = (e) => {
  e.preventDefault(); // <- Stops Original anchor/Link navigation action ❌ 
  router.push(e.currentTarget.href); // <- Uses Routing instead 🍕🍕🍕 
  setOpenSearch(false); // <- Close SearchDialog afterwards
};

<Link
  href={`/user/${login}`}
  ....
  onClick={handleClick}
>
  <Avatar size="sm" className="!rounded-full flex-none" avatarURL={getAvatarByUsername(login)} />
  <div className="flex items-center gap-2 overflow-hidden">
    .....
  </div>
</Link>
```
🍕

OgDev-01 commented on interfaces/global-state-types.ts:
```diff
@@ -10,4 +10,5 @@ export interface GlobalStateInterface {
   providerToken?: string | null;
   userId?: number | null;
   hasReports?: boolean;
+  openSearch?: boolean;
```
Curious why this is global?

babblebey replied:
Great question haha 😅

It was getting somewhat uneasy to manage the `openSearch` state, as I have started to drill down the `setOpenSearch` dispatch function almost 4 components deep to the search results list item.

State Tree

- AuthSection (Declares the `openSearch` state)
  - SearchDialogTrigger (uses the `setOpenSearch`)
  - SearchDialog (receives the `setOpenSearch` as prop from AuthSection to close the dialog onClick/keyboard shortcuts)
    - SearchResult (receives the `setOpenSearch` as prop from SearchDialog to pass it down to its child component`
      - UserResultCard - (receives the `setOpenSearch` as prop from `SearchResult` to close dialog onClick/keyboardCtrl result item.

OgDev-01 replied:
Nice! 👍. Looks good to me

nickytonline commented on lib/hooks/useDebounceTerm.ts:
```diff
@@ -0,0 +1,18 @@
+
+const useDebounceTerm = (value: string, delay = 500) => {
+  const [debouncedValue, setDebouncedValue] = useState("");
+  const timeoutRef = useRef();
+
+  useEffect(() => {
+    timeoutRef.current = setTimeout(() => setDebouncedValue(value), delay) as unknown as typeof timeoutRef.current;
```
Refactor (blocking): use the native type.

```suggestion
  const timeoutRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    timeoutRef.current = setTimeout(() => setDebouncedValue(value), delay);
```

nickytonline commented on components/organisms/TopNav/top-nav.tsx:
```diff
@@ -18,43 +33,30 @@ const TopNav: React.FC = () => {
-        <AuthSection />
-      </div>
-    </header>
+    <nav className={className}>
```
Refactor (blocking): There are more than one `<nav />` elements on the page. To help screen readers, [add an aria-label for the nav element](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/navigation_role#associated_wai-aria_roles_states_and_properties). Insights seem to make sense for the name, but open to suggestions. See 

```suggestion
    <nav className={className} aria-label="Insights">
```

babblebey replied:
Yea, its something around `main`

nickytonline commented on components/molecules/AuthSection/auth-section.tsx:
```diff
@@ -92,50 +96,56 @@ const AuthSection: React.FC = ({}) => {
+                  <Divider type="vertical" className="!h-6 !bg-light-slate-6"></Divider>
+                </>
+              ) : (
+                ""
```
Refactor: `null` is prefered when not returning anything in JSX.

```suggestion
                null
```

or you can do `!onboarded && your_JSX`

babblebey replied:
It was outside the scope of the PR at first!

But, I'll appropriate 🤞

babblebey commented:
🍕

brandonroberts reviewed (approved):
🍕
